src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -563,5 +623,45 @@ void RayletBasedActorScheduler::HandleWorkerLeaseReply(
   }
 }
 
+void GcsActorScheduler::HandleWorkerLeaseRejectedReply(
+    std::shared_ptr<GcsActor> actor, const rpc::RequestWorkerLeaseReply &reply) {
+  // The request was rejected because of insufficient resources.
+  if (!actor->GetAcquiredResources().IsEmpty()) {
+    // Return the actor's acquired resources, which updates GCS' resource view.
+    ReturnActorAcquiredResources(actor);
+  }
+  if (normal_task_resources_changed_callback_ &&
+      RayConfig::instance().gcs_actor_scheduling_enabled()) {
+    normal_task_resources_changed_callback_(actor->GetNodeID(), reply.resources_data());
+  }
+  actor->UpdateAddress(rpc::Address());
+  Reschedule(actor);
+}
+
+void GcsActorScheduler::OnActorDestruction(std::shared_ptr<GcsActor> actor) {
+  if (!actor->GetAcquiredResources().IsEmpty()) {
+    ReturnActorAcquiredResources(actor);
+    cluster_task_manager_->ScheduleAndDispatchTasks();
+  }
+}
+
+void GcsActorScheduler::ReturnActorAcquiredResources(std::shared_ptr<GcsActor> actor) {
+  auto &cluster_resource_manager =
+      cluster_task_manager_->GetClusterResourceScheduler()->GetClusterResourceManager();
+  cluster_resource_manager.AddNodeAvailableResources(
+      scheduling::NodeID(actor->GetNodeID().Binary()), actor->GetAcquiredResources());
+  actor->SetAcquiredResources(ResourceRequest());
+}
+
+size_t GcsActorScheduler::GetPendingActorsCount() const {
+  return cluster_task_manager_->GetInfeasibleQueueSize() +
+         cluster_task_manager_->GetWaitingQueueSize();
+}
+
+bool GcsActorScheduler::RemovePendingActor(const std::shared_ptr<GcsActor> &actor) {
+  return cluster_task_manager_->CancelTask(
```

Comment:
This interface is very confusing. Can you tell me why we need this (and the scenario where it is used?)

Reply:
Yes, this function does mean to cancel an in-flight actor scheduling.